design-proposals/klip-43-timestamp-data-type-support.md

```diff
@@ -0,0 +1,159 @@
+# KLIP-43: TIMESTAMP Data Type Support
+
+**Author**: @jzaralim | 
+**Release Target**: 0.15 | 
+**Status**: In Discussion | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/6649
+
+**tl;dr:** _KSQL currently converts time data into long values. This KLIP introduces a TIMESTAMP
+data type to handle time data instead._
+
+## Motivation and background
+
+With Connect integration, KSQL processes data from various sources and sinks. Most other databases
+have a TIMESTAMP or DATETIME type, which KSQL currently converts into long values. This sometimes
+causes problems when sinking into a database. Supporting TIMESTAMP types would make moving time data
+between KSQL and other data sources/sinks smoother and less error-prone.
+
+## What is in scope
+### Required
+
+* Add TIMESTAMP type to KSQL
+* Serialization and de-serialization of TIMESTAMPs to Avro, JSON and Delimited formats
+* Built-in functions to support TIMESTAMP usage
+
+### Future enhancements
+
+* Give window units (HOUR, DAY etc) a TIMESTAMP value
+* Add DATE and TIME types
+* UDFs such as DAY and HOUR
+
+## What is not in scope
+Changing the ROWTIME data type. We will eventually want this to happen, but that is a separate discussion.
+
+## Public APIS
+
+The TIMESTAMP data type will store a time without timezone information. It will be useable in any
+place STRING or INTEGER could be used:
+```roomsql
+CREATE STREAM stream_name (col1 TIMESTAMP, COL2 STRING) AS ...
+CREATE TABLE table_name (col1 STRUCT<field TIMESTAMP>) AS ...
+```
+
+## Design
+
+### Serialization/Deserialization
+
+TIMESTAMPs will be handled by the Java Date type within KSQL. The corresponding Kafka Connect type is
+[org.apache.kafka.connect.data.Timestamp](https://kafka.apache.org/0100/javadoc/org/apache/kafka/connect/data/Timestamp.html).
+They are represented as long types in Schema Registry, but also come with a tag indicating that it is a timestamp, so they should be distinguishable from long types when handling serialized values in KSQL.
+
+#### Avro
+
+Avro schemas represent timestamps as
+```json
+{
+  "type": "long",
+  "logicalType": "timestamp-millis"
+}
+```
+The Avro deserializer KSQL uses supports this.
+
+#### JSON/Delimited
+
+Timestamps will get stored in JSON and CSV files as long values. The KSQL JSON and delimited deserializers
+will be updated to parse timestamps.
+
+### UDFs
+
+The following UDFs should be updated to use the TIMESTAMP type instead of BIGINT:
```

Comment:
Will BIGINT be discontinued? If so, that might break compatibility, right? Or will there an implicity casting of BIGINT -> TIMESTAMP?

For instance, I wonder what would happen if these scenarios:
- TIMESTAMPTOSTRING(timestamp_col, 'yyyy-MM-dd HH:mm:ss zzz','America/Los_Angeles')
- TIMESTAMPTOSTRING(bigint_col, 'yyyy-MM-dd HH:mm:ss zzz','America/Los_Angeles')
- TIMESTAMPTOSTRING(cast(int_col as bigint), 'yyyy-MM-dd HH:mm:ss zzz','America/Los_Angeles')
^ Will this fail because we're casting to BIGINT?

Reply:
BIGINT/TIMESTAMP implicit casting makes sense. In that case, all three of those commands would work